core/drivers/bcm_sotp.c

```diff
@@ -111,25 +128,155 @@ TEE_Result bcm_iproc_sotp_mem_read(uint32_t row_addr, uint32_t sotp_add_ecc,
 	/* No ECC check till SOTP_NO_ECC_ROWS */
 	if (row_addr > SOTP_NO_ECC_ROWS &&
 	    reg_val & SOTP_STATUS_1__ECC_DET) {
-		EMSG("SOTP ECC ERROR Detected ROW %d\n", row_addr);
+		EMSG("SOTP ECC ERROR Detected ROW %d", row_addr);
 		read_data = SOTP_ECC_ERR_DETECT;
 	}
 
 	/* Command done is cleared */
 	io_setbits32((bcm_sotp_base + SOTP_STATUS_1), SOTP_STATUS_1__CMD_DONE);
 	io_clrbits32((bcm_sotp_base + SOTP_PROG_CONTROL),
 		     SOTP_PROG_CONTROL__OTP_CPU_MODE_EN);
-	DMSG("read done\n");
+	DMSG("read done");
 
 	*rdata = read_data;
 	return ret;
 }
 
+TEE_Result bcm_iproc_sotp_mem_write(uint32_t row_addr, bool sotp_add_ecc,
+				    uint64_t wdata)
+{
+	uint32_t chip_state = 0;
+	uint32_t chip_ctrl_default = 0;
+	uint32_t chip_ctrl = 0;
+	uint8_t prog_array[4] = { 0x0F, 0x04, 0x08, 0x0D };
+	TEE_Result ret = TEE_SUCCESS;
+
+	assert(bcm_sotp_base);
+
+	chip_state = io_read32(bcm_sotp_base + SOTP_REGS_SOTP_CHIP_STATES);
+
+	if (chip_state & CHIP_STATE_DEFAULT) {
+		chip_ctrl_default = io_read32(bcm_sotp_base + SOTP_CHIP_CTRL);
+		DMSG("SOTP: enable special prog mode");
+
+		chip_ctrl = BIT(SOTP_CHIP_CTRL_SW_OVERRIDE_CHIP_STATES) |
+			    BIT(SOTP_CHIP_CTRL_SW_MANU_PROG) |
+			    BIT(SOTP_CHIP_CTRL_SW_CID_PROG) |
+			    BIT(SOTP_CHIP_CTRL_SW_AB_DEVICE);
+
+		io_write32(bcm_sotp_base + SOTP_CHIP_CTRL, chip_ctrl);
+	}
+
+	/* Check for FDONE status */
+	ret = otp_status_done_wait(bcm_sotp_base + SOTP_STAT_0,
+				   SOTP_STATUS_0__FDONE);
+	if (ret) {
+		EMSG("FDONE status done wait failed and returned %X", ret);
```

Comment:
`ret` is based on a `uint32_t` so it's better formatted with `PRIx32`, I'd suggest:
``` C
"... %#"PRIx32, ret);
```
to make it clear that it's in hexadecimal.

Reply:
```c
		EMSG("FDONE status done wait failed and returned %#"PRIx32, ret);
```